proton_driver/dbapi/cursor.py

```diff
@@ -368,3 +457,12 @@ def _check_cursor_closed(self):
     def _check_query_started(self):
         if self._state == self._states.NONE:
             raise ProgrammingError('no results to fetch')
+
+    def _get_cols_and_rows(self):
+        cols = self._columns
+        rows = self._rows
+        if 'tuple_cast' in self._columns[0]:
```

Comment:
when will the 'tuple_cast' occur? can you give some example to explain?

Reply:
Also, need check len(_columns) > 0 to avoid exception  when no columns is return in query

Reply:
1. When execute `select (a, b, c) from test`, db will retuen a type `tuple_cast(a, b, c)` and a rows `((123, 'abc', datetime.datetime(2024, 5, 20, 12, 11, 10)),)`.At first I thought a,b,c should be parsed from tuple_cast,but I think it should not be parsed.

But when there is exactly one column in the table and `columnar == False`,proton-driver will parse every record as a tuple with a None like `(1,)`.Is it a bug?I think the answer is `(1)` to be correct.Should I fix it or just add a check here?

2. Every query executed by cursor with the options `with_column_types==True`. And construct a datafrme need colunm header,if user execute DDL or other commands without a columns header they shouldn't call these functions.

Reply:
It is better to always check boundary before access container for safety. User may use the API in an unexpected way and/or bugs may occur in future change.

BTW, I saw in client, it converts space in column name to underscore. Is this needed in dbapi?

https://github.com/timeplus-io/proton-python-driver/blob/develop/proton_driver/client.py#L409